// @shopify/koa-shopify-webhooks/build/node/register.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registerWebhook = registerWebhook;
exports.DeliveryMethod = exports.ApiVersion = void 0;

var _network = require("@shopify/network");

var _types = require("./types");

let ApiVersion;
exports.ApiVersion = ApiVersion;

(function (ApiVersion) {
  ApiVersion["April19"] = "2019-04";
  ApiVersion["July19"] = "2019-07";
  ApiVersion["October19"] = "2019-10";
  ApiVersion["January20"] = "2020-01";
  ApiVersion["April20"] = "2020-04";
  ApiVersion["July20"] = "2020-07";
  ApiVersion["October20"] = "2020-10";
  ApiVersion["Unstable"] = "unstable";
  ApiVersion["Unversioned"] = "unversioned";
})(ApiVersion || (exports.ApiVersion = ApiVersion = {}));

let DeliveryMethod;
exports.DeliveryMethod = DeliveryMethod;

(function (DeliveryMethod) {
  DeliveryMethod["Http"] = "http";
  DeliveryMethod["EventBridge"] = "eventbridge";
})(DeliveryMethod || (exports.DeliveryMethod = DeliveryMethod = {}));

async function registerWebhook({
  address,
  topic,
  accessToken,
  shop,
  apiVersion,
  deliveryMethod = DeliveryMethod.Http
}) {
  const response = await fetch(`https://${shop}/admin/api/${apiVersion}/graphql.json`, {
    method: _network.Method.Post,
    body: buildQuery(topic, address, deliveryMethod),
    headers: {
      [_types.WebhookHeader.AccessToken]: accessToken,
      [_network.Header.ContentType]: 'application/graphql'
    }
  });
  const result = await response.json();
  return {
    success: isSuccess(result, deliveryMethod),
    result
  };
}

function isSuccess(result, deliveryMethod) {
  switch (deliveryMethod) {
    case DeliveryMethod.Http:
      return Boolean(result.data && result.data.webhookSubscriptionCreate && result.data.webhookSubscriptionCreate.webhookSubscription);

    case DeliveryMethod.EventBridge:
      return Boolean(result.data && result.data.eventBridgeWebhookSubscriptionCreate && result.data.eventBridgeWebhookSubscriptionCreate.webhookSubscription);
  }
}

function buildQuery(topic, address, deliveryMethod) {
  let mutationName;
  let webhookSubscriptionArgs;

  switch (deliveryMethod) {
    case DeliveryMethod.Http:
      mutationName = 'webhookSubscriptionCreate';
      webhookSubscriptionArgs = `{callbackUrl: "${address}"}`;
      break;

    case DeliveryMethod.EventBridge:
      mutationName = 'eventBridgeWebhookSubscriptionCreate';
      webhookSubscriptionArgs = `{arn: "${address}"}`;
      break;
  }

  return `
    mutation webhookSubscriptionCreate {
      ${mutationName}(topic: ${topic}, webhookSubscription: ${webhookSubscriptionArgs}) {
        userErrors {
          field
          message
        }
        webhookSubscription {
          id
        }
      }
    }
  `;
}
