// @shopify/dates/dist/src/parse-date-string-parts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/**
 * Allowed date string formats
 * yyyy-mm-dd
 * yyyy-mm-ddThh:mm:ss.fff
 * yyyy-mm-ddThh:mm:ss.fff+hh:mm
 * yyyy-mm-ddThh:mm:ss.fff-hh:mm
 * yyyy-mm-ddThh:mm:ss.fffZ
 */
var DATE_TIME_PARTS_REGEX = /^(\d{4})-(\d{2})-(\d{2})(?:T(\d{2}):(\d{2}):(\d{2})(?:\.(\d{1,3}))?(?:(Z|(?:(\+|-)(\d{2}):(\d{2}))))?)?$/;
function parseDateStringParts(dateString) {
    var dateTimeParts = new RegExp(DATE_TIME_PARTS_REGEX).exec(dateString);
    if (dateTimeParts == null) {
        return null;
    }
    // slice the first regex part (the full match) off
    var _a = tslib_1.__read(Array.from(dateTimeParts).slice(1), 11), year = _a[0], month = _a[1], day = _a[2], hour = _a[3], minute = _a[4], second = _a[5], millisecond = _a[6], timeZoneOffset = _a[7], sign = _a[8], timeZoneHour = _a[9], timeZoneMinute = _a[10];
    return {
        year: year,
        month: month,
        day: day,
        hour: hour,
        minute: minute,
        second: second,
        millisecond: millisecond,
        timeZoneOffset: timeZoneOffset,
        sign: sign,
        timeZoneHour: timeZoneHour,
        timeZoneMinute: timeZoneMinute,
    };
}
exports.parseDateStringParts = parseDateStringParts;
